src/main/scala/tech/sourced/gemini/Gemini.scala

```diff
@@ -49,81 +52,182 @@ class Gemini(session: SparkSession) {
     save(files)
   }
 
+  /**
+    * Search for duplicates and similar items to the given one.
+    *
+    * @param inPath path to an item
+    * @param conn   Database connection
+    * @return
+    */
+  def query(inPath: String, conn: Session): ReportByLine = {
+    val path = new File(inPath)
+    if (path.isDirectory) {
+      ReportByLine(Gemini.findDuplicateProjects(path, conn, keyspace))
+      //TODO: implement based on Apolo
+      //findSimilarProjects(path)
+    } else {
+      ReportByLine(Gemini.findDuplicateItemForFile(path, conn, keyspace))
+      //TODO: implement based on Apolo
+      //findSimilarFiles(path)
+    }
+  }
+
+  /**
+    * Finds duplicate files among hashed repositories
+    * It is used one query per distinct file
+    *
+    * @param conn Database connections
+    * @return
+    */
+  def report(conn: Session): ReportExpandedGroup = {
+    ReportExpandedGroup(Gemini.findAllDuplicateItems(conn, keyspace))
+  }
+
+  /**
+    * Finds duplicate files among hashed repositories
+    * It is used only one query
+    * (Only supported by Apache Cassandra databases)
+    *
+    * @param conn Database connections
+    * @return
+    */
+  def reportCassandraCondensed(conn: Session): ReportGrouped = {
+    ReportGrouped(Gemini.findAllDuplicateBlobHashes(conn, keyspace))
+  }
+
+  /**
+    * Finds duplicate files among hashed repositories
+    * It is used one query per unique duplicate file, plus an extra one
+    * (Only supported by Apache Cassandra databases)
+    *
+    * @param conn Database connections
+    * @return
+    */
+  def reportCassandraGroupBy(conn: Session): ReportExpandedGroup = {
+    val duplicates = reportCassandraCondensed(conn).v
+      .map { item =>
+        Gemini.findDuplicateItemForBlobHash(item.sha, conn, keyspace)
+      }
+    ReportExpandedGroup(duplicates)
+  }
+
+  def applySchema(session: Session): Unit = {
+    println("CQL: creating schema") //TODO(bzz): Log.Debug
+    Source
+      .fromFile(Gemini.defaultSchemaFile)
+      .getLines
+      .map(_.trim)
+      .filter(!_.isEmpty)
+      .foreach { line =>
+        val cql = line.replace("__KEYSPACE__", keyspace)
```

Comment:
is it possible to just execute `USE $keyspace` before CQL and remove prefix in cql file?

Reply:
It could be one possible further improvement. 
For now, this replacement already looks awesome.

Reply:
I'm not sure if it could cause unexpected problems. Could it happen something weird if running parallel tests with a shared connection, or would it be forbidden?